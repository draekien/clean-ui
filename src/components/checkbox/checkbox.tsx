// eslint-disable-next-line tsdoc/syntax
/** @jsxImportSource theme-ui */
import React from 'react';
import { Size } from '../../types/layouts';

export interface CheckboxProps {
  size?: Size;
  checked?: boolean;
  disabled?: boolean;
  onClick?: (checked?: boolean) => void;
}

const getScale = (size: Size) => {
  switch (size) {
    case 'small':
      return {
        box: '1.25rem',
        checkmark: {
          height: '0.5rem',
          width: '0.125rem',
          top: 'calc(50% - 0.4rem)',
          left: 'calc(50% - 0.125rem)',
        },
      };
    case 'large':
      return {
        box: '2rem',
        checkmark: {
          height: '0.75rem',
          width: '0.25rem',
          top: 'calc(50% - 0.5rem)',
          left: 'calc(50% - 0.25rem)',
        },
      };
    default:
      return {
        box: '1.5rem',
        checkmark: {
          height: '0.75rem',
          width: '0.25rem',
          top: 'calc(50% - 0.5rem)',
          left: 'calc(50% - 0.25rem)',
        },
      };
  }
};

export const Checkbox: React.FC<CheckboxProps> = ({
  size = 'medium',
  checked,
  disabled,
  onClick,
}) => {
  const handleClick = () => {
    onClick && onClick(!checked);
  };

  const { box, checkmark } = getScale(size);

  return (
    <button
      sx={{
        display: 'inline-block',
        height: box,
        width: box,
        position: 'relative',
        transition: 'all 100ms',
        borderRadius: 'sm',
        border: '2px solid',
        borderColor: 'border',
        backgroundColor: checked ? 'p-400' : 'background',
        cursor: disabled ? 'not-allowed' : 'pointer',
        ':after': {
          top: checkmark.top,
          left: checkmark.left,
          content: '""',
          position: 'absolute',
          transition: 'all 100ms',
          opacity: checked ? 1 : 0,
          height: checkmark.height,
          width: checkmark.width,
          border: 'solid white',
          borderWidth: '0 3px 3px 0',
          transform: 'rotate(45deg)',
        },
        ':hover:not(:disabled), :focus': {
          backgroundColor: checked ? 'p-300' : 'background-lighter',
        },
      }}
      onClick={handleClick}
      disabled={disabled}
    />
  );
};
