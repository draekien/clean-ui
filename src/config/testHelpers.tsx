import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
