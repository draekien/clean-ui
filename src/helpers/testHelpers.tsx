import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
