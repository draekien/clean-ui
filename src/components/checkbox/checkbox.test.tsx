import React from 'react';
import { Checkbox } from './checkbox';
import { themedRender } from '../../helpers/testHelpers';

describe('Checkbox component', () => {
  it('should match snapshot', () => {
    const { container } = themedRender(<Checkbox />);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when checked', () => {
    const { container } = themedRender(<Checkbox checked />);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when disabled', () => {
    const { container } = themedRender(<Checkbox disabled />);

    expect(container).toMatchSnapshot();
  });
});
