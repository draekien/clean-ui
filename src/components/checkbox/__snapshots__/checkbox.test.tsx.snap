// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should match snapshot 1`] = `
<div>
  <button
    class="css-1ca21km-Checkbox"
  />
</div>
`;

exports[`Checkbox component should match snapshot when checked 1`] = `
<div>
  <button
    class="css-zy4xe3-Checkbox"
  />
</div>
`;

exports[`Checkbox component should match snapshot when disabled 1`] = `
<div>
  <button
    class="css-jqmbgc-Checkbox"
    disabled=""
  />
</div>
`;
